File: src/main/kotlin/xyz/xenondevs/renderer/model/UnresolvedModel.kt
package xyz.xenondevs.renderer.model

import com.google.gson.JsonObject
import org.joml.Vector3f
import xyz.xenondevs.renderer.model.cache.TextureCache
import xyz.xenondevs.renderer.model.resource.ResourceId
import xyz.xenondevs.renderer.model.resource.ResourceLoader
import xyz.xenondevs.renderer.util.getAllDoubles
import xyz.xenondevs.renderer.util.getAllFloats
import xyz.xenondevs.renderer.util.getBoolean
import xyz.xenondevs.renderer.util.getFloat
import xyz.xenondevs.renderer.util.getInt
import xyz.xenondevs.renderer.util.getOrNull
import xyz.xenondevs.renderer.util.getString
import xyz.xenondevs.renderer.vector.Vectors
import xyz.xenondevs.renderer.vector.divided
import xyz.xenondevs.renderer.vector.toFloatArray
import java.awt.image.BufferedImage

private fun JsonObject.getVector3f(name: String): Vector3f {
    val values = getAsJsonArray(name).getAllFloats()
    return Vector3f(values[0], values[1], values[2])
}

private fun JsonObject.getVector3fOrNull(name: String): Vector3f? {
    val values = getOrNull(name)?.asJsonArray?.getAllFloats() ?: return null
    return Vector3f(values[0], values[1], values[2])
}

internal class UnresolvedModel(id: ResourceId, val loader: ResourceLoader, obj: JsonObject) {
    
    private var layered = false
    private val parent: UnresolvedModel? = obj.getString("parent")?.let {
        if (ResourceId.of(it).toString() != "minecraft:builtin/generated") {
            return@let loader.modelCache.get(it)
        } else {
            layered = true
            return@let null
        }
    } ?: if (id.path.startsWith("block/") && id.path != "block/block") loader.modelCache.get("block/block") else null
    
    private val textureNames: Map<String, String> = obj.getOrNull("textures")?.asJsonObject?.entrySet()?.associate { it.key to it.value.asString } ?: emptyMap()
    private val elements: List<UnresolvedElement> = obj.getOrNull("elements")?.asJsonArray?.map { UnresolvedElement(this, it.asJsonObject) } ?: emptyList()
    
    val ambientOcclusion: Boolean?
    val rotation: Vector3f?
    val translation: Vector3f?
    val scale: Vector3f?
    
    init {
        val guiObject = obj.getOrNull("display")?.asJsonObject?.getOrNull("gui")?.asJsonObject
        rotation = guiObject?.getVector3fOrNull("rotation")
        translation = guiObject?.getVector3fOrNull("translation")
        scale = guiObject?.getVector3fOrNull("scale")
        ambientOcclusion = obj.getBoolean("ambientocclusion")
    }
    
    @Suppress("NAME_SHADOWING")
    fun resolve(): Model {
        var elements: List<UnresolvedElement>? = null
        var textures = HashMap<String, String>()
        
        var ambientOcclusion: Boolean? = null
        var rotation: Vector3f? = this.rotation
        var translation: Vector3f? = this.translation
        var scale: Vector3f? = this.scale
        
        var layered = false
        var current: UnresolvedModel? = this
        while (current != null) {
            if (elements == null) {
                if (current.elements.isNotEmpty())
                    elements = current.elements
                
            }
            current.textureNames.forEach { (key, value) ->
                if (key !in textures) textures[key] = value
            }
            
            if (rotation == null && translation == null && scale == null) {
                rotation = current.rotation
                translation = current.translation
                scale = current.scale
            }
            
            if (ambientOcclusion == null)
                ambientOcclusion = current.ambientOcclusion
            
            layered = current.layered
            current = current.parent
        }
        
        if (layered) {
            val layers = ArrayList<BufferedImage>()
            
            // negative layers are ignored, layer2 is ignored if there is no layer1 
            var i = 0
            while (true) {
                val texture = textureNames["layer$i"] ?: break
                layers += loader.textureCache.get(texture, 0, 0.0, 0.0, 1.0, 1.0)
                i++
            }
            
            return LayeredModel(layers)
        } else {
            textures = textures.entries.associateTo(HashMap()) { (key, value) ->
                var value = value
                while (value.startsWith('#')) {
                    value = textures[value.substring(1)] ?: break
                }
                
                key to value
            }
            
            val resolvedElements = elements?.map { it.resolve(textures) } ?: emptyList()
            return GeometricalModel(
                resolvedElements,
                ambientOcclusion ?: true,
                rotation ?: Vectors.ZERO,
                translation ?: Vectors.ZERO,
                scale ?: Vectors.ONE
            )
        }
    }
    
}

internal class UnresolvedElement(val model: UnresolvedModel, obj: JsonObject) {
    
    val from: Vector3f = obj.getVector3f("from").divided(16f)
    val to: Vector3f = obj.getVector3f("to").divided(16f)
    
    private val rotation: ElementRotation?
    
    private val faces: Map<Direction, UnresolvedTexture> = obj.getOrNull("faces")?.asJsonObject?.entrySet()
        ?.associate {
            val direction = Direction.valueOf(it.key.uppercase())
            return@associate direction to UnresolvedTexture(this@UnresolvedElement, direction, it.value.asJsonObject)
        } ?: emptyMap()
    
    init {
        rotation = obj.getOrNull("rotation")?.asJsonObject?.let {
            ElementRotation(
                it.getVector3f("origin").divided(16f),
                Axis.valueOf(it.getString("axis")!!.uppercase()),
                it.getFloat("angle")!!,
                it.getBoolean("rescale", false)
            )
        }
        
    }
    
    fun resolve(textures: Map<String, String>): Element {
        val faceTextures = faces.mapValues { it.value.resolve(textures) }.toMutableMap()
        Direction.values().forEach {
            if (!faceTextures.containsKey(it))
                faceTextures[it] = TextureCache.EMPTY_TEXTURE
        }
        
        return Element(
            from,
            to,
            rotation,
            faceTextures
        )
    }
    
}

internal class UnresolvedTexture(val element: UnresolvedElement, direction: Direction, obj: JsonObject) {
    
    private val texture: String = obj.getString("texture")!!.removePrefix("#")
    private val rotation: Int = obj.getInt("rotation") ?: 0
    private val fromX: Double
    private val fromY: Double
    private val toX: Double
    private val toY: Double
    
    init {
        val uv = obj.getOrNull("uv")?.asJsonArray?.getAllDoubles()
        if (uv != null) {
            fromX = uv[0] / 16.0
            fromY = uv[1] / 16.0
            toX = uv[2] / 16.0
            toY = uv[3] / 16.0
        } else {
            val dynamicUV = getDynamicUV(element, direction)
            fromX = dynamicUV[0]
            fromY = dynamicUV[1]
            toX = dynamicUV[2]
            toY = dynamicUV[3]
        }
        
    }
    
    fun resolve(textures: Map<String, String>): BufferedImage {
        return element.model.loader.textureCache.get(
            textures[texture] ?: throw IllegalStateException("Undefined texture: $texture. Available: $textures"),
            rotation,
            fromX, fromY,
            toX, toY
        )
    }
    
    private fun getDynamicUV(parent: UnresolvedElement, direction: Direction): DoubleArray {
        val fromPos = parent.from.toFloatArray().map { it - 0.5 }
        val toPos = parent.to.toFloatArray().map { it - 0.5 }
        
        val axisHor: Int // horizontal axis
        val axisVert: Int // vertical axis
        
        when (direction.axis) {
            Axis.X -> { // west or east
                axisHor = 2 // z
                axisVert = 1 // y
            }
            
            Axis.Y -> { // up or down
                axisHor = 0 // x
                axisVert = 2 // z
            }
            
            Axis.Z -> { // north or south
                axisHor = 0 // x
                axisVert = 1 // y
            }
        }
        
        var nx = -1
        var ny = -1
        if (direction == Direction.WEST || direction == Direction.SOUTH) {
            nx = 1
        } else if (direction == Direction.UP) {
            ny = 1
            nx = 1
        }
        
        val x = doubleArrayOf(fromPos[axisHor] * nx, toPos[axisHor] * nx).sorted()
        val y = doubleArrayOf(fromPos[axisVert] * ny, toPos[axisVert] * ny).sorted()
        
        return doubleArrayOf(x[0], y[0], x[1], y[1]).map { it + 0.5 }.toDoubleArray()
    }
    
}

File: renderer-rs/src/model/unresolved_model.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::ops::Div;

use jni::JNIEnv;
use serde_json::{to_vec, Value};

use crate::model::direction::{Axis, Direction, ElementRotation};
use crate::model::resource::resource_id::ResourceId;
use crate::model::resource::resource_manager::ResourceManager;
use crate::util::math::Vector3d;

fn as_vector(json: &Value) -> Vector3d {
    as_vector_safe(json).unwrap_or_else(|| panic!("Invalid vector: {}", json))
}

fn as_vector_safe(json: &Value) -> Option<Vector3d> {
    let array = json.as_array()?;
    if array.len() == 3 {
        let x = array[0].as_f64()?;
        let y = array[1].as_f64()?;
        let z = array[2].as_f64()?;
        return Some(Vector3d::new(x, y, z));
    }
    None
}

#[derive(Clone)]
pub struct UnresolvedModel {
    layered: bool,
    parent: Option<Box<UnresolvedModel>>,
    texture_names: HashMap<String, String>,
    elements: Vec<UnresolvedElement>,
    ambient_occlusion: bool,
    rotation: Option<Vector3d>,
    translation: Option<Vector3d>,
    scale: Option<Vector3d>,
}

impl UnresolvedModel {
    pub fn new(env: &mut JNIEnv, manager: &mut ResourceManager, id: ResourceId, json: &Value) -> UnresolvedModel {
        let (parent, layered) = Self::find_parent(env, manager, &id, json);
        let textures = Self::parse_textures(json);
        let elements: Vec<UnresolvedElement> = Vec::new();

        let mut rotation = None;
        let mut translation = None;
        let mut scale = None;
        let mut ambient_occlusion = true;

        if let Some(gui_json) = json["display"]["gui"].as_object() {
            if let Some(rotation_json) = gui_json.get("rotation") {
                rotation = as_vector_safe(rotation_json)
            }
            if let Some(translation_json) = gui_json.get("translation") {
                translation = as_vector_safe(translation_json)
            }
            if let Some(scale_json) = gui_json.get("scale") {
                scale = as_vector_safe(scale_json)
            }
            if let Some(ambient_occlusion_json) = gui_json.get("ambientocclusion") {
                ambient_occlusion = ambient_occlusion_json.as_bool().unwrap_or(true);
            }
        }

        let mut model = UnresolvedModel {
            layered,
            parent: parent.map(|model| Box::new(model)),
            texture_names: textures,
            elements,
            ambient_occlusion,
            rotation,
            translation,
            scale,
        };

        Self::parse_elements(&mut model, json);
        model
    }

    fn find_parent<'a>(env: &mut JNIEnv, manager: &'a mut ResourceManager, id: &ResourceId, json: &Value) -> (Option<UnresolvedModel>, bool) {
        return if let Some(parent_name) = json["parent"].as_str() {
            let parent_id = ResourceId::of(parent_name)
                .unwrap_or_else(|_| panic!("Invalid parent id: {}", parent_name));

            if parent_id.to_string() != "minecraft:builtin/generated" {
                let model = manager.get_model(env, &parent_id);
                (Some(model), false)
            } else { (None, true) }
        } else {
            if id.path.starts_with("block/") && id.path != "block/block" {
                let model = manager.get_model(env, &ResourceId::of("minecraft:block/block").unwrap());
                (Some(model), false)
            } else { (None, false) }
        };
    }

    fn parse_textures(json: &Value) -> HashMap<String, String> {
        let mut textures = HashMap::new();
        if let Some(obj) = json["textures"].as_object() {
            for (key, value) in obj {
                if let Some(value) = value.as_str() {
                    textures.insert(key.to_string(), value.to_string());
                }
            }
        }
        textures.remove("particle");
        textures
    }

    fn parse_elements(model: &mut UnresolvedModel, json: &Value) {
        if let Some(array) = json["elements"].as_array() {
            for element_json in array {
                model.elements.push(UnresolvedElement::new(element_json));
            }
        }
    }
}

#[derive(Clone)]
pub struct UnresolvedElement {
    from: Vector3d,
    to: Vector3d,
    rotation: Option<ElementRotation>,
    faces: HashMap<Direction, UnresolvedTexture>,
}

impl UnresolvedElement {
    pub fn new(json: &Value) -> UnresolvedElement {
        let from = as_vector(&json["from"]) / 16.0;
        let to = as_vector(&json["to"]) / 16.0;
        let rotation =
            if let Some(rotation_json) = json["rotation"].as_object() {
                Some(ElementRotation {
                    origin: as_vector(&rotation_json["origin"]) / 16.0,
                    axis: rotation_json["axis"].to_string().to_uppercase().parse::<Axis>()
                        .unwrap_or_else(|_| panic!("Invalid axis: {}", rotation_json["axis"])),
                    angle: rotation_json["angle"].as_f64()
                        .unwrap_or_else(|| panic!("Invalid angle: {}", rotation_json["angle"])),
                    rescale: rotation_json["rescale"].as_bool().unwrap_or(false),
                })
            } else { None };
        let faces = HashMap::new();

        let mut element = UnresolvedElement { from, to, rotation, faces };
        if let Some(faces_json) = json["faces"].as_object() {
            for (key, value) in faces_json {
                let direction = key.to_string().to_uppercase().parse::<Direction>()
                    .unwrap_or_else(|_| panic!("Invalid direction: {}", key));
                let texture = UnresolvedTexture::new(&element, direction.clone(), value);
                element.faces.insert(direction, texture);
            }
        }

        element
    }
}

#[derive(Clone)]
pub struct UnresolvedTexture {
    pub texture_name: String,
    pub rotation: i32,
    pub from_x: f64,
    pub from_y: f64,
    pub to_x: f64,
    pub to_y: f64,
}

impl UnresolvedTexture {
    pub fn new(element: &UnresolvedElement, direction: Direction, json: &Value) -> UnresolvedTexture {
        let texture_name = json["texture"].as_str()
            .unwrap_or_else(|| panic!("Invalid texture: {}", json["texture"]))
            .trim_start_matches('#')
            .to_string();

        let rotation = json["rotation"].as_i64().unwrap_or(0) as i32;

        let (from_x, from_y, to_x, to_y) =
            if let Some(uvs_json) = json["uv"].as_array() {
                let uvs = uvs_json.iter().map(|v| v.as_f64().unwrap() / 16.0).collect::<Vec<f64>>();
                (uvs[0], uvs[1], uvs[2], uvs[3])
            } else {
                Self::get_dynamic_uv(element, direction.clone())
            };

        UnresolvedTexture { texture_name, rotation, from_x, from_y, to_x, to_y }
    }

    fn get_dynamic_uv(parent: &UnresolvedElement, direction: Direction) -> (f64, f64, f64, f64) {
        let from_pos = Vector3d::new(parent.from.x - 0.5, parent.from.y - 0.5, parent.from.z - 0.5);
        let to_pos = Vector3d::new(parent.to.x - 0.5, parent.to.y - 0.5, parent.to.z - 0.5);

        let (axis_hor, axis_vert) = match direction.axis() {
            Axis::X => (2, 1), // west or east -> z, y
            Axis::Y => (0, 2), // up or down -> x, z
            Axis::Z => (0, 1) // north or south -> x, y
        };

        let (nx, ny) = if direction == Direction::WEST || direction == Direction::SOUTH {
            (1.0, -1.0)
        } else if direction == Direction::UP {
            (1.0, 1.0)
        } else {
            (-1.0, 1.0)
        };

        let mut x = [from_pos[axis_hor] * nx, to_pos[axis_hor] * nx];
        x.sort_by(|a, b| a.partial_cmp(b).unwrap());
        let mut y = [from_pos[axis_vert] * ny, to_pos[axis_vert] * ny];
        y.sort_by(|a, b| a.partial_cmp(b).unwrap());

        (x[0] + 0.5, y[0] + 0.5, x[1] + 0.5, y[1] + 0.5)
    }
}